### PR TITLE
Added card auth sign serial and challenge handlers PRF-4731

### PR DESCRIPTION
**Pull request by amanCypherock**
https://app.clickup.com/t/9002019994/PRF-4731

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-4731 Card Authentication: Refactor](https://app.clickup.com/t/9002019994/PRF-4731)

**Review by kartikCypherock — APPROVED** [review]
Conditionally approving the change, based on integration testing with SDK/Cysync

File: apps/manager_app/auth_card.c [hunk 1]
```diff
@@ -507,9 +549,9 @@ void card_auth_handler(manager_query_t *query) {
   ASSERT(NULL != query);
 
   /* Validate if this flow is allowed */
-  if (false == onboarding_step_allowed(ONBOARDING_CARD_AUTHENTICATION)) {
-    // TODO: Reject query
-  }
+  // if (false == onboarding_step_allowed(ONBOARDING_CARD_AUTHENTICATION)) {
+  //   // TODO: Reject query
+  // }
```
Comment (amanCypherock) [inline apps/manager_app/auth_card.c:554]:
```suggestion
  if (false == onboarding_step_allowed(ONBOARDING_CARD_AUTHENTICATION)) {
    // TODO: Reject query
  }
```

File: apps/manager_app/auth_card.c [hunk 2]
```diff
@@ -111,6 +112,7 @@ typedef struct {
   char heading[30];
   char message[100];
   uint8_t acceptable_cards;
+  uint8_t family_id[FAMILY_ID_SIZE];
   bool pair_card_required;
 } auth_card_screen_ctx_t;
```
Comment (kartikCypherock) [inline apps/manager_app/auth_card.c:117]: Is this structure a duplicate of `card_sign_data_config_t` ?

File: apps/manager_app/auth_card.c [hunk 3]
```diff
@@ -332,20 +346,33 @@ static manager_error_code_t handle_sign_card_serial(
...
+  card_error_type_e status = card_sign_auth_data(&sign_serial_config);
+
+  switch (status) {
+    case CARD_OPERATION_SUCCESS:
+      memcpy(resp->serial_signature.signature,
+             sign_serial_config.signature,
+             sizeof(resp->serial_signature.signature));
+      if (DEFAULT_UINT32_IN_FLASH == U32_READ_BE_ARRAY(get_family_id())) {
+        set_family_id_flash(auth_card_data->ctx.family_id);
+      }
```
Comment (kartikCypherock) [inline apps/manager_app/auth_card.c:365]: We are already writing family ID in flash memory during "card tap check".. so are we deeming it unsafe to be there?

File: apps/manager_app/auth_card.c [hunk 4]
```diff
@@ -539,6 +581,9 @@ void card_auth_handler(manager_query_t *query) {
     }
 
     send_auth_card_response(&resp);
+    if (MANAGER_AUTH_CARD_RESPONSE_COMMON_ERROR_TAG == resp.which_response) {
```
Comment (kartikCypherock) [inline apps/manager_app/auth_card.c:584]: Please consider re-visiting this flow again (when time permits) to improve handling of exit conditions...

File: apps/manager_app/auth_card.c (hunk 3 above)
Comment (amanCypherock) [reply apps/manager_app/auth_card.c:365]: removed

File: apps/manager_app/auth_card.c (hunk 2 above)
Comment (amanCypherock) [reply apps/manager_app/auth_card.c:117]: `card_sign_data_config_t` takes a pointer and the `auth_card_screen_ctx_t` object member is passed to that config.

This is helpful when FF value is passed in family id.

File: apps/manager_app/auth_card.c (hunk 4 above)
Comment (amanCypherock) [reply apps/manager_app/auth_card.c:581]: Agreed, I understand that sending response and errors from here is not ideal, so we can update this as follows
- Each handler should send their own response to host for success case 
- Send error response to host in errors cases before showing error display.
- The parent should only be concerned with correct exit condition i.e. flow complete or error occurred
- Flow complete should be indicated by flow status.
- For early exits, say auth failed after the serial sign, handler can return MANAGER_TASK_FAILED to exit.

Comment (ujjwal-cyph) [reply apps/manager_app/auth_card.c:581]: can add these points as TODO here

Comment (amanCypherock) [reply apps/manager_app/auth_card.c:581]: will do.

Comment (kartikCypherock) [reply apps/manager_app/auth_card.c:581]: Got it, please add as to do so we can move ahead

File: apps/manager_app/auth_card.c [hunk 5]
```diff
@@ -324,6 +335,9 @@ static void prepare_card_auth_context(auth_card_data_t *auth_card_data) {
            sizeof(auth_card_data->ctx.message),
            UI_TEXT_PLACE_CARD_TILL_BEEP,
            SIGN_SERIAL_BEEP_COUNT(auth_card_data->ctx.pair_card_required));
+
+  memcpy(auth_card_data->ctx.family_id, get_family_id(), FAMILY_ID_SIZE);
+  return MANAGER_TASK_SUCCESS;
 }
 
 static manager_error_code_t handle_sign_card_serial(
```
Comment (ujjwal-cyph) [inline apps/manager_app/auth_card.c:343]: body of `handle_sign_card_serial` and `handle_sign_challenge` is almost same. would be best if the duplication is removed.

Comment (amanCypherock) [reply apps/manager_app/auth_card.c]: updated, please check

File: apps/manager_app/auth_card.c [hunk 6]
```diff
@@ -324,28 +335,76 @@ static void prepare_card_auth_context(auth_card_data_t *auth_card_data) {
            sizeof(auth_card_data->ctx.message),
            UI_TEXT_PLACE_CARD_TILL_BEEP,
            SIGN_SERIAL_BEEP_COUNT(auth_card_data->ctx.pair_card_required));
+
+  memcpy(auth_card_data->ctx.family_id, get_family_id(), FAMILY_ID_SIZE);
+  return MANAGER_TASK_SUCCESS;
 }
 
-static manager_error_code_t handle_sign_card_serial(
+static manager_error_code_t handle_sign_data(
```
[the remark this answers is not in the document]
Comment (amanCypherock) [reply apps/manager_app/auth_card.c:363]: Added

**Review by ujjwal-cyph — APPROVED** [review]

**Pull request merged**